### PR TITLE
Invertible generators

### PR DESCRIPTION
**Pull request by NickHu**
Work-in-progress

Tracking issue: #238

### CODE REVIEW

**Comment by NickHu** [comment]
dfa7e5f closes #236 
3fa1d69 closes #241
f06f013 closes #244 and #245

**Comment by doctorn** [comment]
(Trying to cancel an oriented invertible scalar with it's inverse between singular slices now causes an `Option::unwrap` to panic)

**Comment by calintat** [comment]
> (Trying to cancel an oriented invertible scalar with it's inverse between singular slices now causes an `Option::unwrap` to panic)

What do you mean by "between singular slices"?

**Comment by doctorn** [comment]
> > (Trying to cancel an oriented invertible scalar with it's inverse between singular slices now causes an `Option::unwrap` to panic)
> 
> What do you mean by "between singular slices"?

Sorry -- I don't know the proper terminology. Just cancelling their vertical composition rather than their horizontal composition.

**Comment by calintat** [comment]
> > > (Trying to cancel an oriented invertible scalar with it's inverse between singular slices now causes an `Option::unwrap` to panic)
> > 
> > 
> > What do you mean by "between singular slices"?
> 
> Sorry -- I don't know the proper terminology. Just cancelling their vertical composition rather than their horizontal composition.

Right that's what I thought you meant. Weird I don't get a panic if I do it, but it just gives the wrong result? Where is the panic thrown?

**Comment by doctorn** [comment]
> > > > (Trying to cancel an oriented invertible scalar with it's inverse between singular slices now causes an `Option::unwrap` to panic)
> > > 
> > > 
> > > What do you mean by "between singular slices"?
> > 
> > 
> > Sorry -- I don't know the proper terminology. Just cancelling their vertical composition rather than their horizontal composition.
> 
> Right that's what I thought you meant. Weird I don't get a panic if I do it, but it just gives the wrong result? Where is the panic thrown?

```
panicked at 'called `Option::unwrap()` on a `None` value', homotopy-core/src/rewrite.rs:1072:74

Stack:

getImports/imports.wbg.__wbg_new_693216e109162396@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web.js:383:21
console_error_panic_hook::hook::h3442c3630cd546fa@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[1665]:0x21001f
core::ops::function::Fn::call::hb65e39c43681a605@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[4473]:0x28737e
std::panicking::rust_panic_with_hook::h84feca33bd4bd229@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[2834]:0x268bf8
std::panicking::begin_panic_handler::{{closure}}::hd2eacd3bb9ff1eab@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[3118]:0x2751ad
std::sys_common::backtrace::__rust_end_short_backtrace::h976699518d897fb1@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[4076]:0x285ffd
rust_begin_unwind@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[3647]:0x28236d
core::panicking::panic_fmt::hc171d095bc4a492d@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[3652]:0x28249c
core::panicking::panic::h3b7a8a1eaf47fc0f@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[3585]:0x28136d
std::thread::local::LocalKey<T>::with::hf33143e9e74aa882@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[1124]:0x1c8bbd
homotopy_core::rewrite::RewriteN::compose::h3923a5a40978f8ca@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[240]:0x6c2e4
homotopy_core::check::<impl homotopy_core::rewrite::Cone>::check::h97b3f981759836ea@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[269]:0x965c3
homotopy_core::check::<impl homotopy_core::rewrite::RewriteN>::check_worker::hb04fe81206cc962b@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[621]:0x1569e1
homotopy_core::rewrite::RewriteN::new::hf4e3ab4ae175e86b@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[1279]:0x1e096f
homotopy_core::contraction::contract_base::ha8d2e2f46664523c@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[261]:0x8d25e
homotopy_core::contraction::contract_in_path::h629c82ef469ae0f3@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[316]:0xc7888
homotopy_core::contraction::<impl homotopy_core::diagram::DiagramN>::contract::{{closure}}::he72b2161ea23b75d@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[1561]:0x20483b
homotopy_core::attach::attach_worker::h7429f6bb4681de1c@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[413]:0x1068c7
homotopy_web::model::proof::ProofState::homotopy_contraction::h781bbe6c529375f6@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[380]:0xf4d46
homotopy_web::model::proof::ProofState::update::h3a9418cdc38049e6@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[205]:0x16f1d
homotopy_web::model::State::update::h33c4b9c67d722e9d@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[212]:0x30eb6
<homotopy_web::app::App as yew::html::component::Component>::update::h7f84b0bd259ebc68@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[411]:0x105831
<yew::html::component::lifecycle::UpdateRunner<COMP> as yew::scheduler::Runnable>::run::h4e20840a4ba4fa42@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[579]:0x149570
std::thread::local::LocalKey<T>::with::h20f9d2963f4b335f@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[2022]:0x23244f
yew::scheduler::start::h5f2fec2fcc41133e@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[4472]:0x287376
<homotopy_web::app::diagram_svg::DiagramSvg<_> as yew::html::component::Component>::view::{{closure}}::h9742ae2273454687@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[2981]:0x26fb78
yew::callback::Callback<IN>::emit::h1e86cb0f3daf1e12@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[2973]:0x26f5be
<yew::html::listener::events::onauxclick::Wrapper as yew::virtual_dom::listeners::Listener>::handle::h0a25ea7124fe1105@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[4369]:0x286fc4
yew::virtual_dom::listeners::Registry::run_handlers::{{closure}}::hd04453a2b1b90afb@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[1145]:0x1cc17f
yew::virtual_dom::listeners::Registry::handle::hb6909d507abf8ef8@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[1949]:0x22bf34
yew::virtual_dom::listeners::GlobalHandlers::ensure_handled::{{closure}}::{{closure}}::h9608686493699d35@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[4118]:0x2862e9
<dyn core::ops::function::Fn<(A,)>+Output = R as wasm_bindgen::closure::WasmClosure>::describe::invoke::hd23c73f3232d1972@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web_bg.wasm:wasm-function[3904]:0x284ed2
__wbg_adapter_38@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web.js:262:10
real@https://homotopy-rs--homotopy-io-labelled-xvbfeqnk.web.app/homotopy_web.js:244:20
```

File: homotopy-core/src/factorization.rs [hunk 1]
```diff
@@ -208,19 +207,15 @@ impl Iterator for ConeIterator {
...
-                self.source.clone(),
-                self.target.clone(),
-                slices.iter().step_by(2).cloned().collect(),
-                slices.into_iter().skip(1).step_by(2).collect(),
-            );
-            if cone.check(Mode::Shallow).is_ok() {
-                return Some(cone);
-            }
-        }
+        let slices = self.slices_product.next()?;
```
Comment (calintat) [inline homotopy-core/src/factorization.rs:210]: I don't think this is correct, `check` is the well-formedness check not typechecking.

File: homotopy-core/src/rewrite.rs [hunk 2]
```diff
@@ -1005,8 +1003,8 @@ impl Cone {
         index: usize,
         source: Vec<Cospan>,
         target: Cospan,
-        regular_slices: Vec<Rewrite>,
-        singular_slices: Vec<Rewrite>,
+        regular_slices: &[Rewrite],
+        singular_slices: &[Rewrite],
```
Comment (calintat) [inline homotopy-core/src/rewrite.rs:1007]: Could you please revert this change? It's not necessary after my change and it's causing a merge conflict.

File: homotopy-model/src/proof.rs [hunk 3]
```diff
@@ -990,25 +1077,42 @@ impl Default for RenderStyle {
     }
 }
 
-#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize)]
 pub struct AttachOption {
     pub generator: Generator,
+    pub diagram: DiagramN,
+    pub tag: Option<&'static str>,
```
Comment (regular-citizen) [inline homotopy-model/src/proof.rs:1084]: This is the bit that is preventing us from deriving Deserialize for AttachOption. We should probably be able to change this to Option<String>, at the price of a few clones.

Comment (calintat) [reply homotopy-model/src/proof.rs:1084]: Go for it.
